draft-ietf-httpbis-unprompted-auth.md

```diff
@@ -205,10 +231,26 @@ of bytes necessary.
 
 ## Key Exporter Output {#output}
 
-The output of the exporter is a 32-byte symmetric key. That symmetric key is
-then used as nonce which can be signed using the client's chosen asymmetric
-private key. The resulting signature is then transmitted to the server using
-the Authorization field.
+The TLS key exporter output is described in {{fig-output}}:
```

Comment:
I think that you want to be clearer about what is going on here.  In prose.  That is, 48 bytes is output, then the first 32 bytes is input to the signature and the next 16 bytes are sent along with the signature (to allow the recipient to confirm that the exporter produces the right values).

@dennisjackson, could you just transmit half the nonce instead of extending the KDF output?  The other half should be hard to guess, but confirming that way would save an extra invocation of HMAC.  (I'm not particularly concerned about the cost; I'm just asking to improve my understanding.)

Reply:
Why not just send the entire nonce?

Reply:
If we send the entire nonce, then a lazy implementer could use that as the value for signature checking instead of properly computing the key exporter. The current design (or MT's proposal to only send part of the nonce) avoid that risk.

My gut feeling is that using another part of the key exporter would make security analysis easier than using part of the nonce, but I'll let @dennisjackson comment on that

Reply:
I'm not sure I buy the lazy implementer argument, but I also think sending half the nonce would be sufficient here.

Reply:
I agree that sending half the nonce would work, but what's the benefit? We need to transmit some sort of verification bit regardless, so the only benefit I see to reusing the start of the nonce instead of getting some more key exporter output is that it saves an HKDP operation. I'd rather take the minimal CPU hit to avoid bit reuse if there's no other benefit

Reply:
It just seems like the most obvious thing to do: send the message that is being signed.

Reply:
Sorry for the slow response. 

Short answer: @chris-wood is right that from a formal analysis perspective sending the entire message or a hash of the entire message would also be fine and this is the generic fix to this kind of signature attack. @martinthomson's suggestion of sending half the signature input also works in this specific case. However, for pragmatic reasons I think @DavidSchinazi's solution is nicest overall. 

Longer answer: 

From the perspective of a formal analysis, using the full or half signature input is fine. The positive property we're looking for is that the sender is committed to a single message input, so the full message obviously achieves that. The half input is only going to match with prob 2^128 so also fine. Thinking about negative consequences, we're going from transmitting the signature of the authenticator to the authenticator itself. In a formal analysis, this isn't ever going to be a problem because signature schemes are assumed to be unforgeable (EUF-CMA) but nothing more, so an attacker in 'proof world' can do things like take a signature and reverse it to learn the message (even though that's typically hard in the real world).   

Pragmatically though, I like @DavidSchinazi's solution best. It lessens the risk of a lazy implementer using the sender's value by mistake in the signature verification or trying to save a `memcmp` by skipping the second check of the signature input. 

Another *very* minor benefit that in some signature schemes like Ed25519, the attacker can't reverse it in practice as they only learn `h(pk || m)` and in this case `m` is high entropy. So in the case the implementer uses Ed25519, by using the extra KDF output you make it intractable for an attacker who has witnessed an unprompted auth header to extract the authenticator value and make their own valid header with their key.